src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,185 @@
+        {
+            if (OperationUsesPreviewFeatures(context.Operation, requiresPreviewFeaturesSymbols))
+            {
+                context.ReportDiagnostic(context.Operation.CreateDiagnostic(Rule));
+            }
+        }
+
+        private bool OperationUsesPreviewFeatures(IOperation operation, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            ISymbol? symbol = GetOperationSymbol(operation);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            return OperationUsesPreviewFeatures(operation, symbol, requiresPreviewFeaturesSymbols, true);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                _ => null,
+            };
+
+        private bool TryGetOrCachePreviewFeaturesAttributesOnSymbol(IOperation operation, ISymbol symbol, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols, bool checkParents)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out IOperation? existing))
+            {
+                if (checkParents)
+                {
+                    var containingSymbol = symbol.ContainingSymbol;
+                    // Namespaces do not have attributes
+                    while (containingSymbol is INamespaceSymbol)
+                    {
+                        containingSymbol = containingSymbol.ContainingSymbol;
+                    }
+
+                    if (containingSymbol != null)
+                    {
+                        if (TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, containingSymbol, requiresPreviewFeaturesSymbols, checkParents))
+                        {
+                            return true;
+                        }
+                    }
+                }
+                if (symbol is ITypeSymbol typeSymbol)
+                {
+                    ImmutableArray<INamedTypeSymbol> interfaces = typeSymbol.AllInterfaces;
+                    foreach (INamedTypeSymbol? anInterface in interfaces)
+                    {
+                        if (TryGetOrCachePreviewFeaturesAttributesOnSymbol(operation, anInterface, requiresPreviewFeaturesSymbols, checkParents))
+                        {
+                            return true;
+                        }
+                    }
```

Comment:
Any case that could fail if `AllInterfaces` is replaced with `Interfaces`?

My thought here is that looping the direct interfaces will eventually end up checking the parent interfaces when we recursively call the method with `anInterface`.